### DISCUSSION TITLE
Should I store newsletter subscriber emails in Directus?

### DISCUSSION THREAD

**Discussion by niklasgrewe** [directus/directus#2819] [Archive]
Hi,
In the Directus demo I saw that email addresses were stored in the collection "Mailing List". Is there a specific reason and purpose for storing such data in Directus and not in another database?

I'm thinking about whether it makes more sense to store email newsletter subscribers directly in Directus and not in a MongoDB for example. 

How do you see that?

**Comment by benhaynes** [comment]
This is just preference — storing emails in SQL vs NoSQL that is, and just depends on your use-case. The real question would be: does is make more sense to store emails in a CMS or a CRM, and again this depends on usage. Many people actually build out their Directus instance as a full-featured CRM, and our v9 transactional mail endpoint will make newsletter extensions pretty powerful too.

Maybe I misunderstood your question... but the answer seems very tied to context.

**Reply by niklasgrewe** [reply]
@benhaynes thanks for your answer. No, you didn't misunderstand my question at all. I'm already interested in the mail endpoint, now that you mention it. I'm still using Directus v8 at the moment, can I already use email endpoints etc. to send newsletters to my customers? Could I integrate services like Sendgrid or would I have to use Webhooks to send newsletters via an external API server? In other words: What is already possible with Directus to send newsletters and what are your concrete plans for v9?

It would be really cool, of course, if you could soon use Directus as a full CMS/CRM, with statistics, dashboards, newsletter dispatch and much more... 👍 

and one more question... I would like to set up a customer portal on my website, where customers can login, send inquiries, view and change their data etc. Could I also store this customer data in Directus? Would that have any advantages, or would it be a good idea to separate the services and data a bit, at the latest here, since Directus is a CMS first and foremost 😉

**Reply by benhaynes** [reply]
I would say all of this is only really a good option in v9+. Directus will be able to manage email templates, and provide you an API endpoint for managing/sending emails, but you would still need to use SendGrid (or something) for a more robust email server.

We're actually in the process of "pivoting" from CMS to a Database Content Management Platform... so I'd say it's a great option to ingest/store your other data in Directus. You'll be able to use v9 extensions to essentially use it as a dashboard/CRM/BI/PM/whatever tool. And it has the benefit of being completely customizable/extensible... and maintaining control over your data (in pure SQL).

**Reply by niklasgrewe** [reply]
it all sounds really good and I can't wait to see and try everything then. Should I wait until the release of v9 to save my customer data, or can I already save the data there now? 

Could I theoretically also do the authentication via Directus, or would I have to use additional frameworks like Passport etc.?

**Reply by benhaynes** [reply]
Hard to say, you can try the Alpha and see if it works for your needs... but it is pre-release, so beware! :)

No sure about the auth...

**Reply by niklasgrewe** [reply]
ok, if you have already released the alpha, can you say when the beta version will come?

**Reply by benhaynes** [reply]
Unfortunately, no. It's TBD... but likely in 2-3 weeks.

**Reply by niklasgrewe** [reply]
ok cool glad to hear it. How can I install the Alpha version? I'm currently working with Docker. Could I upgrade from alpha to beta then? Do you have anything built in for that?

**Reply by benhaynes** [reply]
There won't be database migrations for all these pre-release versions (we're releasing 1-ish per day), but the schema won't change much, so you'll be able to do a manual upgrade without too much trouble.

Install: https://github.com/directus/next#%EF%B8%8F-installation

```
npx create-directus-project my-project
```

**Reply by rijkvanzanten** [reply]
> Could I theoretically also do the authentication via Directus, or would I have to use additional frameworks like Passport etc.?

Yeah, you can store your project's users as Directus users and use all of Directus' auth flow in your own app 🙂

**Reply by niklasgrewe** [reply]
@rijkvanzanten cool 😎 that means I could configure the authentication so that each customer only gets read and write access to his data? 

What about the Media Gallery? Can I also create private areas for each customer so that they can upload and download files, pictures etc.? Would this also work if I use storage providers like S3?

I currently use the Directus JS SDK. Can I use it to create new tables, insert data etc.? Is every table automatically visible in the Directus App, or how can I imagine that? In other words: What is the most sensible way to use Directus not only as a CMS but also as a CRM without getting confused which data is now website data or customer data? How can I separate the two?

**Reply by rijkvanzanten** [reply]
> that means I could configure the authentication so that each customer only gets read and write access to his data?

Yes, you can set the permissions up to `mine`, which means that the user only has access to data that they've created / are the owner of.

> What about the Media Gallery? Can I also create private areas for each customer so that they can upload and download files, pictures etc.?

Files can be set to `mine` or `role` as well

> Would this also work if I use storage providers like S3?

Yeah! The storage provider is just an abstraction of where the "source file" (original) is stored. All of Directus' features work the same regardless of storage provider.

> I currently use the Directus JS SDK. Can I use it to create new tables, insert data etc.?

Yeah, you can use the full API from the JS SDK as well. Everything in https://docs.directus.io/api/reference.html can be done through the SDK as well, including creating tables etc.

>  Is every table automatically visible in the Directus App, or how can I imagine that?

By default, admin users have access to everything in Directus, so for admin users, created tables are automatically visible. For non-admin users, they have no permissions to tables by default. If you want them to have access to the newly created table, you'd have to give them (at least) read access to that permission before it shows up in the app. You can do this through the API (and thus SDK) as well if you want to automate that.

> What is the most sensible way to use Directus not only as a CMS but also as a CRM without getting confused which data is now website data or customer data? How can I separate the two?

That's a bit of a too-opinionated question for me to answer fully. I don't know enough about your website/business structure to give a proper assessment on that. It all boils down to personal preference. If your CRM records are different from website users, than you can have a "newsletter signup" or "accounts" or whatever makes most sense table for those, and keep them separate from the website users. You can also store everything under directus_users, and use a different role for both to keep them separated.

**Reply by niklasgrewe** [reply]
@rijkvanzanten really awesome, thanks for your answer and your effort for this great flexible solution 👍 it's a real pleasure to work with. I also have a lot of praise for the new design in v9. Your work was worth it again. And considering that this is still an open source project, it is really world class 💯

**Reply by rijkvanzanten** [reply]
Thank you! That means a lot ☺️❤️

**Reply by niklasgrewe** [reply]
@rijkvanzanten very gladly
> Yeah, you can store your project's users as Directus users and use all of Directus' auth flow in your own app 🙂

with the help of the WebAuth API and the release of iOS 14 it is possible to use FaceID and TouchID from web to authenticate user. It is possible to use it with the Directus Auth Flow? possible to combine this and integrate in my own app? 
that would be really nice 👍

**Reply by rijkvanzanten** [reply]
You'll probably want to add your own middleman server for that auth flow, instead of going straight to Directus from the client app 🙂

**Reply by niklasgrewe** [reply]
@rijkvanzanten thanks for clarified answer 👍 yes that in any case, but i can already save my app users as Directus users, right? Only the auth flow is different. To what extent can I integrate my own Auth Flow here and still use the Directus JS SDK in my app? How can I tell the Directus API that someone is already authenticated, when i authenticate user in my own way? Would I have to create my own SSO provider? Could I theoretically also use frameworks like [Cotter](https://cotter.app) or [Passport](https://passportjs.org) as Auth Flow?

I'm sorry for all the questions, but could you be a little more specific?

**Reply by rijkvanzanten** [reply]
> How can I tell the Directus API that someone is already authenticated, when i authenticate user in my own way?

You can pass the static token of the user to the API to use the API as that user. The way you retrieve this static token is up to you. You can have a middleman server that you use to authenticate users to using whatever you want, and then retrieve and continue using the static token for that user that you've verified separately
